Website: Add blog post for 18.0.0 (#547)
apache/arrow-site: _posts/2024-10-16-18.0.0-release.md

the License.  You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
{% endcomment %}
-->

The Apache Arrow team is pleased to announce the 18.0.0 release. This covers
over 3 months of development work and includes [**XXX resolved issues**][1]
on [**YYY distinct commits**][2] from [**ZZZ distinct contributors**][2].
See the [Install Page](https://arrow.apache.org/install/)
to learn how to get the libraries for your platform.

The release notes below are not exhaustive and only expose selected highlights
of the release. Many other bugfixes and improvements have been made: we refer
you to the [complete changelog][3].

## Community

Since the 17.0.0 release, JJJJJ has been invited to be committer.
No new members have joined the Project Management Committee (PMC).

Thanks for your contributions and participation in the project!

## Linux packages notes

> 

Author: @kou any note here or should we remove `Linux packages notes`?

Reviewer: We have a note:

```suggestion

Azure file system is enabled.
```

## C Data Interface notes

> 

Author: @pitrou can you help with those?

Reviewer 2: There is nothing to add about the C Data Interface AFAICT. However, I've pushed notes for C++ and the columnar format.

## Arrow Flight RPC notes

> 

Author: @lidavidm can you help with the Fight notes?

Reviewer 3: ```suggestion

**Flight UCX is deprecated.** We plan to remove this experiment in the next couple of releases.

The Java implementation now transparently handles compressed Arrow data when reading, instead of requiring explicit configuration. (GH-43469)

The Ruby bindings now support implementing DoPut on the server. (GH-43814)
```

## C++ notes

For C++ notes refer to the full changelog.

### Highlights

> 

Author: @pitrou, @zeroshade

### Acero

> 

Author: @westonpace @zanmato1984 can you help with the acero / compute notes for 18.0.0?

Reviewer 4: Done :)

> 

Reviewer 4: ```suggestion

- Enhanced the row-oriented representation by widening the offset type from 32-bit to 64-bit, resolving crashes and data corruption in aggregation and hash join on large datasets due to offset overflow (GH-43495).
- Improved ordered aggregation performance by reducing complexity from `O(n*m)` to `O(n)`, where `n` is the number of rows and `m` the number of segments in the batch (GH-44052).
```

### Compute

### Dataset

### Filesystems

### GPU

### IPC

### Parquet

### Substrait

## C# notes

> 

Author: @CurtHagenlocher @adamreeve can you help with those?

> 

Reviewer 5: ```suggestion
- Partial support has been added for LargeBinary, LargeString and LargeList. The column sizes cannot exceed 2 GB in length. (GH-43266).
- Changes to Flight support were made for better control and compatibility, and to allow Flight Server to be hosted in pre-Kestrel versions of .NET (GH-43907, GH-43672, GH-41347).
- Support has been added for newly-defined types decimal32 and decimal64 (GH-44271).
- The import of sliced arrays through the C Data interface now works correctly. (GH-43267)
```

## Java notes

> 

Author: @lidavidm @vibhatha @danepitkin can any of you help with the Java notes?

Reviewer 3: ```suggestion

**Java 8 is no longer supported.** (GH-38051)

**Gandiva may not work in this release.** For details, please see [GH-43576](https://github.com/apache/arrow/issues/43576).

Basic support for RunEndEncoded was added (GH-39982). The ListView/StringView vector implementations are now more complete, including C Data support (multiple issues).

Several APIs have been updated to accept `long` for addresses in preparation for FFM/large buffer support (GH-43902). We no longer expose `sun.misc.Unsafe` (GH-43479). We no longer ship the `shaded` flight-core JARs (GH-43217).

More options were added to the Dataset ScanOptions API (GH-28866).
```

Reviewer 6: Sorry I was a bit late for the discussion here. What @lidavidm proposed LGTM!

## JavaScript notes

> 

Author: @domoritz @trxcllnt can you help with the JS notes for the release?

Reviewer 7: Just one relevant change this release:

```suggestion
- Accessing individual rows in Tables or Structs should now be more performant ([GH-30863](https://github.com/apache/arrow/issues/30863)).
```

## Python notes

> 

Author: @AlenkaF @jorisvandenbossche can you help with those?

## R notes

> For more on what’s in the 18.0.0 R package, see the [R changelog][4].

Author: @jonkeane @paleolimbot @assignUser can you help with the R notes?

## Ruby and C GLib notes

> 

Author: @kou @mrkn can you help with the Ruby and CGLib notes?

Reviewer: Done!

Reviewer 8: @kou Thanks!!

### Ruby

### C GLib

## Rust notes and Go notes

The Rust and Go projects have moved to separate repositories outside the
main Arrow monorepo. For notes on the latest release of the Rust
implementation, see the latest [Arrow Rust changelog][5].
For notes on the latest release of the Go implementation, see the latest

> [Arrow Go changelog][6]

Author: @zeroshade I've added Go to the notes as we do with Rust, do you want to add something more?

Reviewer 9: I think that's fine for now. Guess I'll have to do my own blog posts for Go releases from now on haha

over 3 months of development work and includes [**XXX resolved issues**][1]
on [**YYY distinct commits**][2] from [**ZZZ distinct contributors**][2].
See the [Install Page](https://arrow.apache.org/install/)
to learn how to get the libraries for your platform.

The release notes below are not exhaustive and only expose selected highlights
of the release. Many other bugfixes and improvements have been made: we refer
you to the [complete changelog][3].

## Community

Since the 17.0.0 release, JJJJJ has been invited to be committer.
No new members have joined the Project Management Committee (PMC).

Thanks for your contributions and participation in the project!

## Linux packages notes

Azure file system is enabled.

## C Data Interface notes

## Arrow Flight RPC notes

## C++ notes

For C++ notes refer to the full changelog.

### Highlights

### Acero

### Compute

### Dataset

### Filesystems

### GPU

### IPC

### Parquet

### Substrait

## C# notes

## Java notes

## JavaScript notes

## Python notes

## R notes

For more on what’s in the 18.0.0 R package, see the [R changelog][4].

## Ruby and C GLib notes

### Ruby

> 

Reviewer: ```suggestion

* Add workaround for install failure due to `re2.pc` on Ubuntu 20.04: [GH-41396](https://github.com/apache/arrow/issues/41396)
* Add support for `0` decimal value: [GH-43877](https://github.com/apache/arrow/issues/43877)

C GLib related improvements are also available in Ruby. 

```

### C GLib

> 

Reviewer: ```suggestion

* Add support for Azure file system: [GH-43738](https://github.com/apache/arrow/issues/43738)
* FlightRPC: Add support for DoPut: [GH-41056](https://github.com/apache/arrow/issues/41056)
* FlightRPC: Add support for timeout: [GH-44178](https://github.com/apache/arrow/issues/44178)
* Parquet: Add support for writing a record batch: [GH-40860](https://github.com/apache/arrow/issues/40860)
* Add support for pull style IPC stream format decoder: [GH-40493](https://github.com/apache/arrow/issues/40493)
```

WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
{% endcomment %}
-->

The Apache Arrow team is pleased to announce the 18.0.0 release. This covers
over 3 months of development work and includes [**XXX resolved issues**][1]
on [**YYY distinct commits**][2] from [**ZZZ distinct contributors**][2].
See the [Install Page](https://arrow.apache.org/install/)
to learn how to get the libraries for your platform.

The release notes below are not exhaustive and only expose selected highlights
of the release. Many other bugfixes and improvements have been made: we refer
you to the [complete changelog][3].

## Community

Since the 17.0.0 release, JJJJJ has been invited to be committer.
No new members have joined the Project Management Committee (PMC).

Thanks for your contributions and participation in the project!

## Columnar format

The Arrow columnar format now allows 32-bit and 64-bit decimal data, in
addition to the already existing 128-bit and 256-bit decimal data types
(GH-43956).

## Linux packages notes

> Azure file system is enabled.

Reviewer 2: Does this section really deserve being at the top?

Reviewer: Ah, we can move this because this is not so important in 18.0.0.
This section was important in 17.0.0 because we dropped support for one platform: https://arrow.apache.org/blog/2024/07/16/17.0.0-release/

## C Data Interface notes

## Arrow Flight RPC notes

> **Flight UCX is deprecated.** We plan to remove this experiment in the next couple of releases.

Reviewer 2: Flight UCX is C++ specific, this shouldn't be at the top of the release notes. The "Arrow Flight RPC notes" should be for Flight specification changes IMHO.

The Java implementation now transparently handles compressed Arrow data when reading, instead of requiring explicit configuration. (GH-43469)

The Ruby bindings now support implementing DoPut on the server. (GH-43814)

## C++ notes

The default memory pool has changed to mimalloc on all platforms (GH-43254).
Previously, jemalloc was used by default on Linux. Using mimalloc by default

> provides a more consistent experience accross different platforms, and

Reviewer 7: typo

```suggestion
provides a more consistent experience across different platforms, and
```

makes configuration easier. It is expected that this might either increase
or decrease performance on user workloads that use the default memory pool;
please benchmark accordingly. Jemalloc can still be selected by setting

> the `ARROW_DEFAULT_MEMORY_POOL` environment variable to "jemalloc".

Reviewer 7: Just linking the env var.

```suggestion
the [`ARROW_DEFAULT_MEMORY_POOL`](https://arrow.apache.org/docs/cpp/env_vars.html#envvar-ARROW_DEFAULT_MEMORY_POOL) environment variable to "jemalloc".
```

The new Decimal32 and Decimal64 types have been made available (GH-43956).

Several canonical extension types have been implemented:
- the [Opaque](https://arrow.apache.org/docs/dev/format/CanonicalExtensions.html#opaque) extension type (GH-43454);
- the [8-bit boolean](https://arrow.apache.org/docs/dev/format/CanonicalExtensions.html#bit-boolean) extension type (GH-17682);
- the [UUID](https://arrow.apache.org/docs/dev/format/CanonicalExtensions.html#uuid) extension type (GH-15058);
- the [JSON](https://arrow.apache.org/docs/dev/format/CanonicalExtensions.html#json) extension type (GH-32538).

### Acero

- Enhanced the row-oriented representation by widening the offset type from 32-bit to 64-bit, resolving crashes and data corruption in aggregation and hash join on large datasets due to offset overflow (GH-43495).
- Improved ordered aggregation performance by reducing complexity from `O(n*m)` to `O(n)`, where `n` is the number of rows and `m` the number of segments in the batch (GH-44052).

### Compute

Casting between string-like and string-view-like types has been implemented (GH-42247).

### Dataset

### Filesystems

> Writing small files to S3 now uses a single S3 API call instead of three

Reviewer 2: Oops, you're right. Do you want to propose a new wording?

> Writing small files to S3 now uses a single S3 API call instead of three

Reviewer 10: Something like:

> Writing smalle files to S3 can uses a single S3 "PutObject" API call instead of "multipart upload" with three api calls when new `sanitize_bucket_on_open_` option off

?

## C# notes

- Partial support has been added for LargeBinary, LargeString and LargeList. The column sizes cannot exceed 2 GB in length. (GH-43266).
- Changes to Flight support were made for better control and compatibility, and to allow Flight Server to be hosted in pre-Kestrel versions of .NET (GH-43907, GH-43672, GH-41347).
- Support has been added for newly-defined types decimal32 and decimal64 (GH-44271).
- The import of sliced arrays through the C Data interface now works correctly. (GH-43267)
## Java notes

**Java 8 is no longer supported.** (GH-38051)

**Gandiva may not work in this release.** For details, please see [GH-43576](https://github.com/apache/arrow/issues/43576).

Basic support for RunEndEncoded was added (GH-39982). The ListView/StringView vector implementations are now more complete, including C Data support (multiple issues).

Several APIs have been updated to accept `long` for addresses in preparation for FFM/large buffer support (GH-43902). We no longer expose `sun.misc.Unsafe` (GH-43479). We no longer ship the `shaded` flight-core JARs (GH-43217).

More options were added to the Dataset ScanOptions API (GH-28866).

## JavaScript notes

## Python notes

> ## R notes
> 
> For more on what’s in the 18.0.0 R package, see the [R changelog][4].

Reviewer 11: ```suggestion
## R notes

* R functions that users write that use functions that Arrow supports in dataset
  queries now can be used in queries too. Previously, only functions that used
  arithmetic operators worked.
  For example, `time_hours <- function(mins) mins / 60` worked,
  but `time_hours_rounded <- function(mins) round(mins / 60)` did not;
  now both work. These are automatic translations rather than true user-defined
  functions (UDFs); for UDFs, see `register_scalar_function()`. [GH-41223](https://github.com/apache/arrow/issues/41223)
* `mutate()` expressions can now include aggregations, such as `x - mean(x)`. [GH-41350](https://github.com/apache/arrow/issues/41350)
* `summarize()` supports more complex expressions, and correctly handles cases
  where column names are reused in expressions. [GH-41223](https://github.com/apache/arrow/issues/41223)
* The `na_matches` argument to the `dplyr::*_join()` functions is now supported.
  This argument controls whether `NA` values are considered equal when joining. [GH-41358](https://github.com/apache/arrow/issues/41358)
* R metadata, stored in the Arrow schema to support round-tripping data between
  R and Arrow/Parquet, is now serialized and deserialized more strictly.
  This makes it safer to load data from files from unknown sources into R data.frames. [GH-41969](https://github.com/apache/arrow/issues/41969)
* Turn on the S3 and ZSTD features by default for macOS. [GH-42210](https://github.com/apache/arrow/issues/42210)
* Fix a bug in our implementation of `pull` on grouped datasets, it now
  returns the expected column. [GH-43172](https://github.com/apache/arrow/issues/43172)
  
For full details of what’s in the 18.0.0 R package, see the [R changelog][4].
```

Reviewer 7: @jonkeane these are the 17.0.0 changes though right?

Nic and I are working on a NEWS.md patch in https://github.com/apache/arrow/issues/44496 and we can copy that in here in a bit.

this work for additional information regarding copyright ownership.
The ASF licenses this file to you under the Apache License, Version 2.0
(the "License"); you may not use this file except in compliance with
the License.  You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
{% endcomment %}
-->

The Apache Arrow team is pleased to announce the 18.0.0 release. This covers
over 3 months of development work and includes [**XXX resolved issues**][1]
on [**YYY distinct commits**][2] from [**ZZZ distinct contributors**][2].
See the [Install Page](https://arrow.apache.org/install/)
to learn how to get the libraries for your platform.

The release notes below are not exhaustive and only expose selected highlights
of the release. Many other bugfixes and improvements have been made: we refer
you to the [complete changelog][3].

## Community

> Since the 17.0.0 release, JJJJJ has been invited to be committer.

Reviewer 10: Should here being updated?

Author: Yes, I will update the number of commits and the new committers once we approve RC0

The new Decimal32 and Decimal64 types have been made available (GH-43956).

Several canonical extension types have been implemented:
- the [Opaque](https://arrow.apache.org/docs/dev/format/CanonicalExtensions.html#opaque) extension type (GH-43454);
- the [8-bit boolean](https://arrow.apache.org/docs/dev/format/CanonicalExtensions.html#bit-boolean) extension type (GH-17682);
- the [UUID](https://arrow.apache.org/docs/dev/format/CanonicalExtensions.html#uuid) extension type (GH-15058);
- the [JSON](https://arrow.apache.org/docs/dev/format/CanonicalExtensions.html#json) extension type (GH-32538).

### Acero

- Enhanced the row-oriented representation by widening the offset type from 32-bit to 64-bit, resolving crashes and data corruption in aggregation and hash join on large datasets due to offset overflow (GH-43495).
- Improved ordered aggregation performance by reducing complexity from `O(n*m)` to `O(n)`, where `n` is the number of rows and `m` the number of segments in the batch (GH-44052).

### Compute

Casting between string-like and string-view-like types has been implemented (GH-42247).

### Dataset

### Filesystems

> Writing small files to S3 now uses a single S3 API call instead of three
> (GH-40557). Files larger than 5 MB still go through the regular multipart
> upload mechanism.

Reviewer 2: ```suggestion
Writing small files to S3 can use a single S3 API call instead of three,
provided the new option `allow_delayed_open` is enabled (GH-40557).
Files larger than 5 MB still go through the regular multipart
upload mechanism.
```

More options were added to the Dataset ScanOptions API (GH-28866).

## JavaScript notes

## Python notes
Compatibility notes:
* NumPy required dependency has been removed from pyarrow packaging
  [GH-43846](https://github.com/apache/arrow/issues/43846) and has been
  made an optional runtime dependency [GH-25118](https://github.com/apache/arrow/issues/25118).
* Support for Python 3.8 has been dropped [GH-43518](https://github.com/apache/arrow/issues/43518)
* No longer used serialize/deserialize Pyarrow C++ functions have been
  deprecated [GH-44063](https://github.com/apache/arrow/issues/44063).
* Passing of build flags to setup.py (e.g. `setup.py --with-parquet`) has been
  deprecated [GH-43514](https://github.com/apache/arrow/issues/43514)

New features:
* Non-cpu work has continued with [GH-43973](https://github.com/apache/arrow/issues/43973),
  [GH-43728](https://github.com/apache/arrow/issues/43728), [GH-43727](https://github.com/apache/arrow/issues/43727),
  [GH-43391](https://github.com/apache/arrow/issues/43391),
  [GH-42222](https://github.com/apache/arrow/issues/42222) and

>   [GH-41665](https://github.com/apache/arrow/issues/41665).

Reviewer 12: I think just enumerating those issue links might not be that informative without more context (for someone reading this, you would have to click on all those links to have an idea what actually changed)
